te2rules/adapter.py

```diff
@@ -229,3 +233,121 @@ def _convert(self) -> DecisionTree:
+
+        tree_ensemble = self._extract_tree_ensemble()
+        self.tree_ensemble = tree_ensemble
+
+        self.random_forest = self._convert()
+
+    def _extract_tree_ensemble(self) -> np.ndarray:
+        """
+        Private method to extract trees
+        from the xgboost.sklearn.XGBClassifier object.
+        """
+
+        # Extract booster and dump it
+        booster = self.xgb_model.get_booster()
+        tree_ensemble = booster.get_dump(fmap="", with_stats=True, dump_format="json")
+        regressors = np.empty(len(tree_ensemble), dtype=object)
+
+        # Check whether feature names are already present in the model
+        self.replace_feature_names = True if booster.feature_names is None else False
+
+        # Iterate over ensemble trees and build them using te2rules.tree.DecisionTree
+        for i, tree in enumerate(tree_ensemble):
+            tree_dict = json.loads(tree)
+
+            # We initiate Decision Tree TreeNode with placeholder values that will be filled later
+            regressor = DecisionTree(TreeNode(node_name="", threshold=1.0))
+
+            self.build_tree(tree_dict, regressor)
+
+            regressors[i] = regressor
+
+        return regressors
+
+    def build_tree(self, node, tree) -> None:
+        """
+        Private method to iteratively build the tree.
+        """
+        if "leaf" in node:
+            tree.node = LeafNode(value=node["leaf"])
+        else:
+            # We initiate left and right Decision Tree TreeNode with placeholder values that will be filled later
+            tree.left = DecisionTree(TreeNode(node_name="", threshold=0.5))
+            tree.right = DecisionTree(TreeNode(node_name="", threshold=0.5))
+
+            # Recursively we build the left / right tree
+            self.build_tree(
+                [
+                    children
+                    for children in node["children"]
+                    if children["nodeid"] == node["yes"]
+                ][0],
+                tree.left,
+            )
+            self.build_tree(
+                [
+                    children
+                    for children in node["children"]
+                    if children["nodeid"] == node["no"]
+                ][0],
+                tree.right,
+            )
+
+            # We define the node_name, depending on whether feature_names are already present in the model or needs to be replaced
+            if self.replace_feature_names:
+                # XGBoost stores variable as f0, f1, f2 etc, therefore we extract in the following way the variable index
+                feature_index = re.search("^f(\d+)", node["split"]).group(1)
+                node_name = self.feature_names[int(feature_index)]
+            else:
+                node_name = node["split"]
+
+            tree.node = TreeNode(
+                node_name=node_name,
+                threshold=(
+                    # TODO: For binary variables threshold is 1, we should have 0.5 instead
```

Comment:
I just checked why this happens. Apparently, scikit and xgboost treat feature splits differently. In scikit, the left child corresponds to feature <= threshold while in xgboost, the left child corresponds to feature < threshold. Note, how they treat equality differently.